// firebase-admin/lib/installations/error.js
/*! firebase-admin v14.1.0 */
"use strict";
/*!
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FirebaseInstallationsError = exports.installationsClientErrorCode = exports.InstallationsErrorCode = void 0;
const error_1 = require("../utils/error");
/**
 * The constant mapping for valid Installations client error codes.
 */
exports.InstallationsErrorCode = {
    INVALID_ARGUMENT: 'invalid-argument',
    INVALID_PROJECT_ID: 'invalid-project-id',
    INVALID_INSTALLATION_ID: 'invalid-installation-id',
    API_ERROR: 'api-error',
};
/**
 * Internal Installations client error code mapping used to construct ErrorInfo.
 */
exports.installationsClientErrorCode = {
    INVALID_ARGUMENT: {
        code: exports.InstallationsErrorCode.INVALID_ARGUMENT,
        message: 'Invalid argument provided.',
    },
    INVALID_PROJECT_ID: {
        code: exports.InstallationsErrorCode.INVALID_PROJECT_ID,
        message: 'Invalid project ID provided.',
    },
    INVALID_INSTALLATION_ID: {
        code: exports.InstallationsErrorCode.INVALID_INSTALLATION_ID,
        message: 'Invalid installation ID provided.',
    },
    API_ERROR: {
        code: exports.InstallationsErrorCode.API_ERROR,
        message: 'Installation ID API call failed.',
    },
};
/**
 * Firebase Installations error code structure. This extends `FirebaseError`.
 */
class FirebaseInstallationsError extends error_1.FirebaseError {
    /**
     *
     * @param info - The error code info.
     * @param message - The error message. This will override the default
     *     message if provided.
     */
    constructor(info, message) {
        // Override default message if custom message provided.
        super({
            code: `installations/${info.code}`,
            message: message || info.message,
            httpResponse: info.httpResponse,
            cause: info.cause,
        });
        /** @internal */
        this.codePrefix = 'installations';
    }
}
exports.FirebaseInstallationsError = FirebaseInstallationsError;
