// firebase-admin/lib/eventarc/error.js
/*! firebase-admin v14.1.0 */
"use strict";
/*!
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FirebaseEventarcError = exports.EventarcErrorCode = void 0;
const error_1 = require("../utils/error");
/**
 * The constant mapping for valid Eventarc client error codes.
 */
exports.EventarcErrorCode = {
    UNKNOWN_ERROR: 'unknown-error',
    INVALID_ARGUMENT: 'invalid-argument',
};
/**
 * Firebase Eventarc error code structure. This extends `FirebaseError`.
 */
class FirebaseEventarcError extends error_1.FirebaseError {
    /**
     * @param info - The error code info.
     * @param message - The error message. If provided, this will override the default message.
     */
    constructor(info, message) {
        super({
            code: `eventarc/${info.code}`,
            message: message || info.message,
            httpResponse: info.httpResponse,
            cause: info.cause,
        });
        /** @internal */
        this.codePrefix = 'eventarc';
    }
}
exports.FirebaseEventarcError = FirebaseEventarcError;
