// firebase-admin/lib/remote-config/error.js
/*! firebase-admin v14.1.0 */
"use strict";
/*!
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FirebaseRemoteConfigError = exports.RemoteConfigErrorCode = exports.ERROR_CODE_MAPPING = void 0;
const error_1 = require("../utils/error");
/** @const {Record<string, RemoteConfigErrorCode>} Remote Config server to client error code mapping. */
exports.ERROR_CODE_MAPPING = {
    ABORTED: 'aborted',
    ALREADY_EXISTS: 'already-exists',
    INVALID_ARGUMENT: 'invalid-argument',
    INTERNAL: 'internal-error',
    FAILED_PRECONDITION: 'failed-precondition',
    NOT_FOUND: 'not-found',
    OUT_OF_RANGE: 'out-of-range',
    PERMISSION_DENIED: 'permission-denied',
    RESOURCE_EXHAUSTED: 'resource-exhausted',
    UNAUTHENTICATED: 'unauthenticated',
    UNKNOWN: 'unknown-error',
};
/**
 * The constant mapping for valid Remote Config client error codes.
 */
exports.RemoteConfigErrorCode = {
    ABORTED: 'aborted',
    ALREADY_EXISTS: 'already-exists',
    FAILED_PRECONDITION: 'failed-precondition',
    INTERNAL_ERROR: 'internal-error',
    INVALID_ARGUMENT: 'invalid-argument',
    NOT_FOUND: 'not-found',
    OUT_OF_RANGE: 'out-of-range',
    PERMISSION_DENIED: 'permission-denied',
    RESOURCE_EXHAUSTED: 'resource-exhausted',
    UNAUTHENTICATED: 'unauthenticated',
    UNKNOWN_ERROR: 'unknown-error',
};
/**
 * Firebase Remote Config error code structure. This extends `FirebaseError`.
 */
class FirebaseRemoteConfigError extends error_1.FirebaseError {
    /**
     * @param info - The error code info.
     * @param message - The error message. If provided, this will override the default message.
     */
    constructor(info, message) {
        super({
            code: `remote-config/${info.code}`,
            message: message || info.message,
            httpResponse: info.httpResponse,
            cause: info.cause,
        });
        /** @internal */
        this.codePrefix = 'remote-config';
    }
}
exports.FirebaseRemoteConfigError = FirebaseRemoteConfigError;
