// firebase-admin/lib/functions/error.js
/*! firebase-admin v14.1.0 */
"use strict";
/*!
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FirebaseFunctionsError = exports.FunctionsErrorCode = exports.FUNCTIONS_ERROR_CODE_MAPPING = void 0;
const error_1 = require("../utils/error");
/** @const {Record<string, FunctionsErrorCode>} Cloud Functions server to client error code mapping. */
exports.FUNCTIONS_ERROR_CODE_MAPPING = {
    ABORTED: 'aborted',
    INVALID_ARGUMENT: 'invalid-argument',
    INVALID_CREDENTIAL: 'invalid-credential',
    INTERNAL: 'internal-error',
    FAILED_PRECONDITION: 'failed-precondition',
    PERMISSION_DENIED: 'permission-denied',
    UNAUTHENTICATED: 'unauthenticated',
    NOT_FOUND: 'not-found',
    UNKNOWN: 'unknown-error',
};
/**
 * The constant mapping for valid Cloud Functions client error codes.
 */
exports.FunctionsErrorCode = {
    ABORTED: 'aborted',
    INVALID_ARGUMENT: 'invalid-argument',
    INVALID_CREDENTIAL: 'invalid-credential',
    INTERNAL_ERROR: 'internal-error',
    FAILED_PRECONDITION: 'failed-precondition',
    PERMISSION_DENIED: 'permission-denied',
    UNAUTHENTICATED: 'unauthenticated',
    NOT_FOUND: 'not-found',
    UNKNOWN_ERROR: 'unknown-error',
    TASK_ALREADY_EXISTS: 'task-already-exists',
};
/**
 * Cloud Functions error code structure. This extends `FirebaseError`.
 */
class FirebaseFunctionsError extends error_1.FirebaseError {
    /**
     * @param info - The error code info.
     * @param message - The error message. If provided, this will override the default message.
     */
    constructor(info, message) {
        super({
            code: `functions/${info.code}`,
            message: message || info.message,
            httpResponse: info.httpResponse,
            cause: info.cause,
        });
        /** @internal */
        this.codePrefix = 'functions';
    }
}
exports.FirebaseFunctionsError = FirebaseFunctionsError;
